File: scripts/preview_web.py
"""Preview the shared dashboard with sample data, without connecting to Twitch.

Run: python scripts/preview_web.py --port 8095
The preview is read-only and only listens on localhost.
"""
from __future__ import annotations

import argparse
import json
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlsplit

ROOT = Path(__file__).resolve().parent.parent


class PreviewHandler(BaseHTTPRequestHandler):
    def do_GET(self) -> None:
        path = urlsplit(self.path).path
        if path in {"/api/history", "/api/csrf"}:
            if path == "/api/csrf":
                payload = {"token": "read-only-preview"}
            else:
                payload = json.loads((ROOT / "tests/fixtures/history.json").read_text())
                query = parse_qs(urlsplit(self.path).query)
                game = query.get("game", [""])[0]
                search = query.get("q", [""])[0].casefold()
                items = [row for row in payload["items"] if
                         (not game or (row["gameId"] or "unknown") == game) and
                         search in f'{row["name"]} {row["gameName"]} {row["campaignName"]}'.casefold()]
                try:
                    offset = max(0, int(query.get("offset", ["0"])[0]))
                except ValueError:
                    offset = 0
                payload.update(items=items[offset:offset + 50], total=len(items), offset=offset, limit=50)
            data = json.dumps(payload).encode()
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(data)))
            self.end_headers()
            self.wfile.write(data)
            return
        if path == "/api/events":
            self.send_response(200)
            self.send_header("Content-Type", "text/event-stream")
            self.send_header("Cache-Control", "no-store")
            self.end_headers()
            try:
                payload = (ROOT / "tests/fixtures/web_state.json").read_text()
                self.wfile.write(f"data: {json.dumps(json.loads(payload))}\n\n".encode())
                self.wfile.flush()
                while True:
                    time.sleep(15)
                    self.wfile.write(b": preview keepalive\n\n")
                    self.wfile.flush()
            except (BrokenPipeError, ConnectionResetError):
                pass
            return
        assets = {"/assets/app.js": "text/javascript", "/assets/theme.js": "text/javascript", "/assets/app.css": "text/css"}
        if path in assets:
            file = ROOT / "web" / path.rsplit("/", 1)[-1]
            content_type = assets[path]
        elif path in ("/", "/campaigns", "/mining", "/settings", "/diagnostics", "/history") or path.startswith("/campaigns/"):
            file = ROOT / "web/index.html"
            content_type = "text/html"
        else:
            self.send_error(404)
            return
        data = file.read_bytes()
        self.send_response(200)
        self.send_header("Content-Type", f"{content_type}; charset=utf-8")
        self.send_header("Cache-Control", "no-store")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def do_POST(self) -> None:
        message = b"This is a read-only preview. Run the miner to use this action."
        self.send_response(409)
        self.send_header("Content-Type", "text/plain")
        self.send_header("Content-Length", str(len(message)))
        self.end_headers()
        self.wfile.write(message)

    do_PUT = do_POST


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--port", type=int, default=8095)
    args = parser.parse_args()
    print(f"Read-only dashboard preview: http://127.0.0.1:{args.port}", flush=True)
    ThreadingHTTPServer(("127.0.0.1", args.port), PreviewHandler).serve_forever()

File: scripts/preview-web.ts
/**
 * Bun-native read-only dashboard preview.
 *
 * Mirrors `scripts/preview_web.py` (which remains the CI-canonical preview)
 * using `Bun.serve()` + `Bun.file()` instead of `http.server`. Loopback-only,
 * no Twitch connection, fictional fixtures only. Writes are rejected with 409
 * so browser suites can verify read-only behavior.
 *
 * Run:
 *   bun run preview            # :8095, same default as preview_web.py
 *   bun scripts/preview-web.ts --port 8095
 *
 * Then: `bun run test:browser` in another terminal.
 */
const REPO = (import.meta.dir ?? process.cwd()).replace(/\/scripts$/, "");

function rootFile(...parts: string[]): string {
  // import.meta.dir is `.../scripts`; fixtures live in `tests/fixtures`, UI in `web/`.
  return [REPO, ...parts].join("/").replace(/\/+/g, "/");
}

function parsePort(): number {
  const flag = process.argv.findIndex((a) => a === "--port" || a === "-p");
  const raw =
    (flag >= 0 ? process.argv[flag + 1] : undefined) ??
    process.env.TDM_PREVIEW_PORT ??
    "8095";
  const port = Number.parseInt(String(raw), 10);
  return Number.isFinite(port) && port > 0 && port < 65536 ? port : 8095;
}

const PORT = parsePort();

const SPA_ROUTES = new Set(["/", "/campaigns", "/mining", "/settings", "/diagnostics", "/history"]);
const ASSETS: Record<string, string> = {
  "/assets/app.js": "text/javascript; charset=utf-8",
  "/assets/theme.js": "text/javascript; charset=utf-8",
  "/assets/app.css": "text/css; charset=utf-8",
};

async function historyPayload(url: URL): Promise<Response> {
  const file = Bun.file(rootFile("tests", "fixtures", "history.json"));
  const payload = (await file.json()) as {
    items: Array<{
      gameId?: string | null;
      name?: string;
      gameName?: string;
      campaignName?: string;
    }>;
  };
  const game = url.searchParams.get("game") ?? "";
  const search = (url.searchParams.get("q") ?? "").toLowerCase();
  const rawOffset = Number.parseInt(url.searchParams.get("offset") ?? "0", 10);
  const offset = Number.isFinite(rawOffset) ? Math.max(0, rawOffset) : 0;
  const items = (payload.items ?? []).filter((row) => {
    const gameOk = !game || (row.gameId || "unknown") === game;
    const haystack = `${row.name ?? ""} ${row.gameName ?? ""} ${row.campaignName ?? ""}`.toLowerCase();
    return gameOk && haystack.includes(search);
  });
  return Response.json(
    { ...payload, items: items.slice(offset, offset + 50), total: items.length, offset, limit: 50 },
    { headers: { "Cache-Control": "no-store" } },
  );
}

function eventsStream(): Response {
  let timer: Timer | undefined;
  const stream = new ReadableStream({
    async start(controller) {
      const encoder = new TextEncoder();
      const state = await Bun.file(rootFile("tests", "fixtures", "web_state.json")).text();
      controller.enqueue(encoder.encode(`data: ${JSON.stringify(JSON.parse(state))}\n\n`));
      timer = setInterval(() => {
        try {
          controller.enqueue(encoder.encode(`: preview keepalive\n\n`));
        } catch {
          // Client went away; cleanup happens in cancel().
        }
      }, 15_000);
    },
    cancel() {
      if (timer !== undefined) clearInterval(timer);
    },
  });
  return new Response(stream, {
    headers: {
      "Content-Type": "text/event-stream",
      "Cache-Control": "no-store",
      Connection: "keep-alive",
    },
  });
}

const server = Bun.serve({
  hostname: "127.0.0.1",
  port: PORT,
  development: process.env.NODE_ENV !== "production",
  async fetch(req): Promise<Response> {
    const url = new URL(req.url);
    const path = url.pathname;

    if (req.method !== "GET" && req.method !== "HEAD") {
      return new Response("This is a read-only preview. Run the miner to use this action.", {
        status: 409,
        headers: { "Content-Type": "text/plain; charset=utf-8" },
      });
    }

    if (path === "/api/csrf") {
      // Fixed placeholder for a loopback-only, read-only preview server;
      // this is not a credential.
      return Response.json({ token: "read-only-preview" });
    }
    if (path === "/api/history") return historyPayload(url);
    if (path === "/api/events") return eventsStream();

    if (path in ASSETS) {
      const fileName = path.split("/").at(-1) ?? "";
      const file = Bun.file(rootFile("web", fileName));
      if (!(await file.exists())) return new Response("Not found", { status: 404 });
      return new Response(file, {
        headers: {
          "Content-Type": ASSETS[path]!,
          "Cache-Control": "no-store",
        },
      });
    }

    if (SPA_ROUTES.has(path) || path.startsWith("/campaigns/")) {
      const file = Bun.file(rootFile("web", "index.html"));
      return new Response(file, {
        headers: { "Content-Type": "text/html; charset=utf-8", "Cache-Control": "no-store" },
      });
    }

    return new Response("Not found", { status: 404 });
  },
});

console.log(`Read-only dashboard preview (Bun): http://127.0.0.1:${server.port}/`);
console.log(`Fixtures: tests/fixtures/web_state.json + history.json (fictional, no live miner)`);
